node_package/src/clientStartup.js

```diff
@@ -128,7 +128,17 @@ function render(el, railsContext) {
 You returned a server side type of react-router error: ${JSON.stringify(reactElementOrRouterResult)}
 You should return a React.Component always for the client side entry point.`);
       } else {
-        ReactDOM.render(reactElementOrRouterResult, domNode);
+        const shouldHydrate = !!(
+          ReactDOM.hydrate &&
+          domNode &&
+          domNode.hasAttribute('data-reactroot')
+        );
```

Comment:
Looks like you mix things here you shouldn't mix. E.g. `!!ReactDOM.hydrate` is enough for `shouldHydrate` decision, but if something is wrong w/ dom node then it's out of scope `ReactDOM.hydrate` vs `ReactDOM.render` problem.

Also, https://github.com/facebook/react/issues/10971